File: src/utils/signature.js
import crypto from 'crypto';

/**
 * 生成爱发电API签名
 * 签名规则: md5(token + params{params}ts{ts}user_id{user_id})
 * 
 * @param {string} token - API Token
 * @param {string} params - JSON字符串格式的参数
 * @param {number} ts - 时间戳（秒级）
 * @param {string} userId - 用户ID
 * @returns {string} MD5签名字符串
 */
export function generateSign(token, params, ts, userId) {
  // 构建签名字符串: token + params{params}ts{ts}user_id{user_id}
  const signString = `${token}params${params}ts${ts}user_id${userId}`;
  
  // 计算MD5哈希
  const sign = crypto.createHash('md5').update(signString).digest('hex');
  
  return sign;
}

/**
 * 生成当前时间戳（秒级）
 * @returns {number} 秒级时间戳
 */
export function generateTimestamp() {
  return Math.floor(Date.now() / 1000);
}

/**
 * 构建API请求参数
 * @param {object} params - 请求参数对象
 * @param {string} userId - 用户ID
 * @param {string} token - API Token
 * @returns {object} 包含签名的完整请求参数
 */
export function buildRequestParams(params, userId, token) {
  const paramsString = JSON.stringify(params);
  const ts = generateTimestamp();
  const sign = generateSign(token, paramsString, ts, userId);

  return {
    user_id: userId,
    params: paramsString,
    ts,
    sign,
  };
}

File: internal/utils/signature.go
package utils

import (
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"time"
)

// GenerateSign 生成爱发电API签名
// 签名规则: md5(token + params{params}ts{ts}user_id{user_id})
func GenerateSign(token string, params string, ts int64, userID string) string {
	signString := fmt.Sprintf("%sparams%sts%duser_id%s", token, params, ts, userID)
	hash := md5.Sum([]byte(signString))
	return hex.EncodeToString(hash[:])
}

// GenerateTimestamp 生成当前时间戳（秒级）
func GenerateTimestamp() int64 {
	return time.Now().Unix()
}

// BuildRequestParams 构建API请求参数
func BuildRequestParams(params interface{}, userID string, token string) (map[string]interface{}, error) {
	if params == nil {
		params = map[string]interface{}{}
	}

	paramsBytes, err := json.Marshal(params)
	if err != nil {
		return nil, fmt.Errorf("params 序列化失败: %w", err)
	}

	paramsString := string(paramsBytes)
	ts := GenerateTimestamp()
	sign := GenerateSign(token, paramsString, ts, userID)

	return map[string]interface{}{
		"user_id": userID,
		"params":  paramsString,
		"ts":      ts,
		"sign":    sign,
	}, nil
}
